### PR TITLE
[KOGITO-4940] move KieBase creation in generated code into a static f…

### PR DESCRIPTION
**Pull request by mariofusco**
…ield

### CODE REVIEW

File: kogito-codegen-modules/kogito-codegen-rules/src/main/java/org/kie/kogito/codegen/rules/ProjectSourceClass.java [hunk 1]
```diff
@@ -42,32 +54,43 @@ public ProjectSourceClass withDependencyInjection(String dependencyInjection) {
...
+        TemplatedGenerator generator = TemplatedGenerator.builder()
+                .withTemplateBasePath(TEMPLATE_RULE_FOLDER)
+                .withPackageName("org.drools.project.model")
+                .withFallbackContext(JavaKogitoBuildContext.CONTEXT_NAME)
+                .build(context, "ProjectRuntime");
+
+        CompilationUnit cu = generator.compilationUnitOrThrow("Could not create CompilationUnit");
+        ClassOrInterfaceDeclaration clazz = cu
+                .findFirst(ClassOrInterfaceDeclaration.class)
+                .orElseThrow(() -> new NoSuchElementException("Compilation unit doesn't contain a class or interface declaration!"));
```
Comment (danielezonca) [inline kogito-codegen-modules/kogito-codegen-rules/src/main/java/org/kie/kogito/codegen/rules/ProjectSourceClass.java:66]:
```suggestion
                .orElseThrow(() -> new InvalidTemplateException(generator, "Compilation unit doesn't contain a class or interface declaration!"));
```

File: kogito-codegen-modules/kogito-codegen-rules/src/main/java/org/kie/kogito/codegen/rules/ProjectSourceClass.java [hunk 2]
```diff
@@ -42,32 +54,43 @@ public ProjectSourceClass withDependencyInjection(String dependencyInjection) {
...
+        toMethods(modelMethod.toKieSessionConfMethod()).forEach(clazz::addMember);
+
+        return cu.toString();
+    }
+
+    private void writeInitKieBasesMethod(ClassOrInterfaceDeclaration clazz) {
+        MethodDeclaration initKieBasesMethod = clazz.findAll(MethodDeclaration.class).stream()
+                .filter(m -> m.getNameAsString().equals("initKieBases"))
+                .findFirst()
+                .orElseThrow(() -> new NoSuchElementException("Cannot find initKieBases method"));
```
Comment (danielezonca) [inline kogito-codegen-modules/kogito-codegen-rules/src/main/java/org/kie/kogito/codegen/rules/ProjectSourceClass.java:85]:
```suggestion
                .orElseThrow(() -> new InvalidTemplateException(generator, "Cannot find initKieBases method"));
```

File: kogito-codegen-modules/kogito-codegen-rules/src/main/java/org/kie/kogito/codegen/rules/ProjectSourceClass.java [hunk 3]
```diff
@@ -15,25 +15,37 @@
...
+import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.ast.body.ClassOrInterfaceDeclaration;
+import com.github.javaparser.ast.body.MethodDeclaration;
+import com.github.javaparser.ast.stmt.BlockStmt;
+import com.github.javaparser.ast.stmt.IfStmt;
+
+import static com.github.javaparser.StaticJavaParser.parse;
+import static org.kie.kogito.codegen.rules.IncrementalRuleCodegen.TEMPLATE_RULE_FOLDER;
 
 public class ProjectSourceClass {
```
Comment (danielezonca) [inline kogito-codegen-modules/kogito-codegen-rules/src/main/java/org/kie/kogito/codegen/rules/ProjectSourceClass.java:36]: Can you rename this class `*Generator` like other generator classes?

File: kogito-codegen-modules/kogito-codegen-rules/src/main/java/org/kie/kogito/codegen/rules/ProjectSourceClass.java [hunk 4]
```diff
@@ -42,32 +54,43 @@ public ProjectSourceClass withDependencyInjection(String dependencyInjection) {
...
+                .build(context, "ProjectRuntime");
+
+        CompilationUnit cu = generator.compilationUnitOrThrow("Could not create CompilationUnit");
+        ClassOrInterfaceDeclaration clazz = cu
+                .findFirst(ClassOrInterfaceDeclaration.class)
+                .orElseThrow(() -> new NoSuchElementException("Compilation unit doesn't contain a class or interface declaration!"));
+
+        if (dependencyInjection != null && dependencyInjection.length() > 0) {
+            clazz.addAnnotation(dependencyInjection);
+        }
```
Comment (danielezonca) [inline kogito-codegen-modules/kogito-codegen-rules/src/main/java/org/kie/kogito/codegen/rules/ProjectSourceClass.java:70]: Please remove `withDependencyInjection` method from this class (and corresponding invocation in `IncrementalRuleCodegen`) and replace this with
```suggestion
        if (context.hasDI()) {
            context.getDependencyInjectionAnnotator().withApplicationComponent(clazz);
        }
```

File: kogito-codegen-modules/kogito-codegen-rules/src/main/java/org/kie/kogito/codegen/rules/ProjectSourceClass.java [hunk 5]
```diff
@@ -42,32 +54,43 @@ public ProjectSourceClass withDependencyInjection(String dependencyInjection) {
...
+        IfStmt ifStmt = initKieBasesMethod.findFirst(IfStmt.class).orElseThrow(() -> new NoSuchElementException());
+        BlockStmt ifBlock = ifStmt.getThenStmt().asBlockStmt();
+        for (String kbaseName : modelMethod.getKieBaseNames()) {
+            ifBlock.addStatement("kbaseMap.put( \"" + kbaseName + "\", " +
+                    "KieBaseBuilder.createKieBaseFromModel( model.getModelsForKieBase( \"" + kbaseName + "\" ), " +
+                    "model.getKieModuleModel().getKieBaseModels().get( \"" + kbaseName + "\" ) ) );\n");
+        }
     }
 
     public void write(MemoryFileSystem srcMfs) {
```
Comment (danielezonca) [inline kogito-codegen-modules/kogito-codegen-rules/src/main/java/org/kie/kogito/codegen/rules/ProjectSourceClass.java:96]: Unused method?

File: kogito-codegen-modules/kogito-codegen-rules/src/main/java/org/kie/kogito/codegen/rules/ProjectSourceClass.java [hunk 6]
```diff
@@ -15,25 +15,37 @@
...
+import com.github.javaparser.ast.stmt.IfStmt;
+
+import static com.github.javaparser.StaticJavaParser.parse;
+import static org.kie.kogito.codegen.rules.IncrementalRuleCodegen.TEMPLATE_RULE_FOLDER;
 
 public class ProjectSourceClass {
 
     public static final String PROJECT_RUNTIME_CLASS = "org.drools.project.model.ProjectRuntime";
     private static final String PROJECT_RUNTIME_RESOURCE_CLASS = PROJECT_RUNTIME_CLASS.replace('.', '/') + ".class";
     private static final String PROJECT_RUNTIME_SOURCE = PROJECT_RUNTIME_CLASS.replace('.', '/') + ".java";
```
Comment (danielezonca) [inline kogito-codegen-modules/kogito-codegen-rules/src/main/java/org/kie/kogito/codegen/rules/ProjectSourceClass.java:40]: You can remove all these constants (and avoid a duplication):
- Move `TemplateGenerator generator` instantiation in the constructor and save it in a field
- Replace `getName()` body with `return generator.generatedFilePath()`

**Review by danielezonca — APPROVED** [review]
Thanks for the PR, approving 👍

**Comment by danielezonca** [comment]
Jenkins run native

**Comment by mariofusco** [comment]
jenkins retest this

**Comment by mariofusco** [comment]
Jenkins run native

**Review by evacchi — APPROVED** [review]

**Pull request merged**